src/TreeDataGrid.tsx

```diff
@@ -0,0 +1,428 @@
+import { forwardRef, useCallback, useMemo } from 'react';
+import type { Key, RefAttributes } from 'react';
+
+import { useLatestFunc } from './hooks';
+import { assertIsValidKeyGetter, isCtrlKeyHeldDown } from './utils';
+import type {
+  CellKeyboardEvent,
+  CellKeyDownArgs,
+  GroupRow,
+  GroupRowHeightArgs,
+  Maybe,
+  Omit,
+  RenderRowProps,
+  RowHeightArgs,
+  RowsChangeData
+} from './types';
+import { renderToggleGroup, SELECT_COLUMN_KEY } from '.';
+import DataGrid from './DataGrid';
+import type { DataGridHandle, DataGridProps } from './DataGrid';
+import { useDefaultRenderers } from './DataGridDefaultRenderersProvider';
+import GroupedRow from './GroupRow';
+import { defaultRenderRow } from './Row';
+
+export interface TreeDataGridProps<R, SR = unknown, K extends Key = Key>
+  extends Omit<DataGridProps<R, SR, K>, 'role' | 'aria-rowcount' | 'rowHeight' | 'onFill'> {
+  rowHeight?: Maybe<number | ((args: GroupRowHeightArgs<R>) => number)>;
+  groupBy: readonly string[];
+  rowGrouper: (rows: readonly R[], columnKey: string) => Record<string, readonly R[]>;
+  expandedGroupIds: ReadonlySet<unknown>;
+  onExpandedGroupIdsChange: (expandedGroupIds: Set<unknown>) => void;
+}
+
+type GroupByDictionary<TRow> = Record<
+  string,
+  {
+    readonly childRows: readonly TRow[];
+    readonly childGroups: readonly TRow[] | Readonly<GroupByDictionary<TRow>>;
+    readonly startRowIndex: number;
+  }
+>;
+
+function TreeDataGrid<R, SR, K extends Key>(
+  {
+    columns: rawColumns,
+    rows: rawRows,
+    rowHeight: rawRowHeight,
+    rowKeyGetter: rawRowKeyGetter,
+    onCellKeyDown: rawOnCellKeyDown,
+    onRowsChange,
+    selectedRows: rawSelectedRows,
+    onSelectedRowsChange: rawOnSelectedRowsChange,
+    renderers,
+    groupBy: rawGroupBy,
+    rowGrouper,
+    expandedGroupIds,
+    onExpandedGroupIdsChange,
+    ...props
+  }: TreeDataGridProps<R, SR, K>,
+  ref: React.Ref<DataGridHandle>
+) {
+  const defaultRenderers = useDefaultRenderers<R, SR>();
+  const rawRenderRow = renderers?.renderRow ?? defaultRenderers?.renderRow ?? defaultRenderRow;
+  const headerAndTopSummaryRowsCount = 1 + (props.topSummaryRows?.length ?? 0);
+  const isRtl = props.direction === 'rtl';
+  const leftKey = isRtl ? 'ArrowRight' : 'ArrowLeft';
+  const rightKey = isRtl ? 'ArrowLeft' : 'ArrowRight';
+  const toggleGroupLatest = useLatestFunc(toggleGroup);
+
+  const { columns, groupBy } = useMemo(() => {
```

Comment:
Extracted from `useCalculatedColumns`. Now we run an extra loop for columns. I don't think it should be an issue.